staging/src/k8s.io/apiserver/pkg/util/proxy/websocket.go

```diff
@@ -0,0 +1,192 @@
+	ctx, err := createWebSocketStreams(req, w, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if ctx.resizeStream != nil {
+		ctx.resizeChan = make(chan remotecommand.TerminalSize)
+		go handleResizeEvents(req.Context(), ctx.resizeStream, ctx.resizeChan)
+	}
+
+	return ctx, nil
+}
+
+// Read terminal resize events off of passed stream and queue into passed channel.
+func handleResizeEvents(ctx context.Context, stream io.Reader, channel chan<- remotecommand.TerminalSize) {
+	defer runtime.HandleCrash()
+	defer close(channel)
+
+	decoder := json.NewDecoder(stream)
+	for {
+		size := remotecommand.TerminalSize{}
+		if err := decoder.Decode(&size); err != nil {
+			break
+		}
+
+		select {
+		case channel <- size:
+		case <-ctx.Done():
+			// To avoid leaking this routine, exit if the http request finishes. This path
+			// would generally be hit if starting the process fails and nothing is started to
+			// ingest these resize events.
+			return
+		}
+	}
+}
+
+// createChannels returns the standard channel types for a shell connection (STDIN 0, STDOUT 1, STDERR 2)
+// along with the approximate duplex value. It also creates the error (3) and resize (4) channels.
+func createChannels(opts Options) []wsstream.ChannelType {
+	// open the requested channels, and always open the error channel
+	channels := make([]wsstream.ChannelType, 5)
+	channels[constants.StreamStdIn] = readChannel(opts.Stdin)
+	channels[constants.StreamStdOut] = writeChannel(opts.Stdout)
+	channels[constants.StreamStdErr] = writeChannel(opts.Stderr)
+	channels[constants.StreamErr] = wsstream.WriteChannel
+	channels[constants.StreamResize] = wsstream.ReadChannel
+	return channels
+}
+
+// readChannel returns wsstream.ReadChannel if real is true, or wsstream.IgnoreChannel.
+func readChannel(real bool) wsstream.ChannelType {
+	if real {
+		return wsstream.ReadChannel
+	}
+	return wsstream.IgnoreChannel
+}
+
+// writeChannel returns wsstream.WriteChannel if real is true, or wsstream.IgnoreChannel.
+func writeChannel(real bool) wsstream.ChannelType {
+	if real {
+		return wsstream.WriteChannel
+	}
+	return wsstream.IgnoreChannel
+}
+
+// createWebSocketStreams returns a "conns" struct containing the websocket connection and
+// streams needed to perform an exec or an attach.
+func createWebSocketStreams(req *http.Request, w http.ResponseWriter, opts Options) (*conns, error) {
+	channels := createChannels(opts)
+	conn := wsstream.NewConn(map[string]wsstream.ChannelProtocolConfig{
+		// WebSocket server only supports remote command version 5.
+		constants.StreamProtocolV5Name: {
+			Binary:   true,
+			Channels: channels,
+		},
+	})
+	conn.SetIdleTimeout(defaultIdleConnectionTimeout)
+	// Opening the connection responds to WebSocket client, negotiating
+	// the WebSocket upgrade connection and the subprotocol.
+	negotiatedProtocol, streams, err := conn.Open(w, req)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("unable to upgrade websocket connection: %v", err))
+		return nil, err
```

Comment:
don't we have to close the connection in all the places where we return errors in this function?

Reply:
Fixed by adding `conn.Close()`

Reply:
I see a lot of defer with conn.Close in the library, but this method is really weird

https://github.com/kubernetes/kubernetes/blob/3ac83f528dde9d6f37f0ca164d5642226f2380a7/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go#L173-L177

It never returns an error , I think that something is odd there, kind of seems it can leak goroutines easily

It creates a websocket servers in a goroutine and waits on a `conn.ready` channel

The `conn.ready` channel is closed ONLY in the `initialize` method

https://github.com/kubernetes/kubernetes/blob/3ac83f528dde9d6f37f0ca164d5642226f2380a7/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go#L211

`initialize` is called from the handle https://github.com/kubernetes/kubernetes/blob/3ac83f528dde9d6f37f0ca164d5642226f2380a7/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go#L246-L248

and it has also another defer `conn.Close()`

Reply:
Agree `conn#Open()` *and* `conn#Close()` block on `<-conn.ready`

As @aojea mentioned, that only unblocks if `conn#initialize()` is called. Is that *guaranteed* to happen or is there any way `websocket.Server{Handshake: conn.handshake, Handler: conn.handle}.ServeHTTP(w, req)` can fail in a way that doesn't call `conn#initialize`, and block the Open()/Close() calls ~forever, leaking the goroutine?

Reply:
I believe I have addressed this issue using a `serverCompleted` channel so waiting on the `ready` channel will not possibly hang. If the select statement returns from the `serverCompleted` channel case, we know the websocket server finished before initialization completed, and we clean up accordingly. After moving one of the `conn#Close()` calls, all the remaining calls are guaranteed to occur *after* the `conn#initialize()` which closes the `ready` channel, making these `conn#Close()` calls safe.

Please have a look and let me know what you think.

Reply:
with the changes to conn#Open(), it now cleans up after itself if it returns an error, and calling Close() here would block on `<-conn.ready`, so drop this call to Close() (the one below still looks good to me)

Reply:
Removed this `conn#Close()` call.